monoio-http-client/src/response.rs

```diff
@@ -15,11 +15,11 @@ pub struct ClientResponse {
     /// The response's extensions
     extensions: Extensions,
     /// Payload
-    body: BoxedFramedPayload,
+    body: B,
```

Comment:
For client, can we use something like box dyn to erase the type information?

Reply:
Won't there be some runtime overhead with box dyn compared to direct method calls? The current trait bound for the body type is `   B: Body<Data = Bytes, Error = HttpError> + From<RecvStream> + From<FramedPayloadRecvr>`. We can change ito just `B: Body<Data = Bytes, Error = HttpError>` by changing the return type to `Response<[HttpBody](https://github.com/monoio-rs/monoio-http/pull/11/files/264ef51ea8cf2117c2b51a69e28b56af66800c9e#diff-31fb2352fb732710926cefcfb93368b8a78ecb35658d67bb187b213895c30a19R26)>`. The request body type is generic, but the response body type is not. That's what hyper seems to be doing.

Reply:
I think using box dyn or using enum should be good for client http response.

1. Users may do request in http or https. In previous implementation, we read the response header and body, then the returned Response does not capture the io, it works fine. But we have to make it work in a stream style to avoid eating too much memory. So, the response must contain io. However, io under http maybe TcpStream(or UnixStream if we support UDS later), under https maybe it is RustlsTlsStream or NativeTlsStream. We cannot return 2 types in the same function. Basicly there are 2 ways to erase these type difference: using enum or using box<dyn>(or write vtable manually), these ways all have similar runtime cost(enum should be slightly better than box due to inline ability).
2. Later, we have to support http encoding, like gzip. They can be implemented as a wrapper of bytes stream(as our Body now), so the returned Response type changes accroading to different encoding headers. We still facing the type unifying problem.
3. Users may want to save the response. If the type can not be written or hard to write, it is not good for them.

> That way the request body type is generic, but the response body type is not.

It's great. I think it is the best solution. Client response body can be a box dyn or enum. But here you change the ClientResponse Body to a generic type, is it as expected?

Reply:
BTW, are you intrested in supporting Content-Encoding?

Reply:
I think what you are saying makes sense. I'll leave it as an enum for now, open to changing to a box dyn in the future.

> It's great. I think it is the best solution. Client response body can be a box dyn or enum. But here you change the ClientResponse Body to a generic type, is it as expected? 

I haven't changed the return type to the enum yet. In my first diff both request and response body were generic. Will fix it to something like below :)

```
impl<B> Client<B>
where
    B: Body<Data = Bytes, Error = HttpError> + 'static,
{
    pub async fn send_request(&self, req: http::Request<B>) -> crate::Result<Response<HttpBody>> {
    }
}
```

Reply:
> BTW, are you intrested in supporting Content-Encoding?

Sure I'll start taking a look at it once this change goes in

Reply:
> I'll leave it as an enum for now, open to changing to a box dyn in the future.

Enum is better than Box in performance. If it can be impl clearly, keep it as enum is better.

Reply:
> Sure I'll start taking a look at it once this change goes in

We can work together by then